src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -142,28 +144,15 @@ protected void handleMessageInternal(final Message<?> message) throws Exception
 				? this.messageKeyExpression.getValue(this.evaluationContext, message)
 				: message.getHeaders().get(KafkaHeaders.MESSAGE_KEY);
 
-		ListenableFuture<?> future;
+		Long timestamp = message.getHeaders().get(KafkaHeaders.TIMESTAMP, Long.class);
```

Comment:
Awesome, will make these changes and get back @artembilan , yes I was thinking about the expression part but was not sure. It does make sense to have a timeExpression, will introduce it and see how it goes.